File: src/main/java/br/com/pnp/compiler/model/construct/Procedure.java
package br.com.pnp.compiler.model.construct;

import br.com.pnp.compiler.model.construct.type.Type;
import br.com.pnp.compiler.model.construct.type.primitives.PrimitiveType;
import br.com.pnp.compiler.model.expression.Expression;
import br.com.pnp.compiler.model.instruction.AssignmentInstruction;
import br.com.pnp.compiler.model.instruction.DeclarationInstruction;
import br.com.pnp.compiler.model.instruction.Instruction;

import java.util.ArrayList;
import java.util.List;

public class Procedure extends Block implements Expression {
    private List<DeclarationInstruction> declarations = new ArrayList<>();
    private Type returnType;
    private List<Variable> input;
    private Variable output;
    final private String name;

    public Procedure(String name) {
        this.input = new ArrayList<>();
        this.output = null;
        this.returnType = PrimitiveType.Nulo;
        this.name = name;
    }

    public void setInput(List<Variable> input) {
        if (this.input.isEmpty()) {
            this.input = input;
        }
    }

    public void setOutput(Variable output) {
        if (this.output == null) {
            this.output = output;
            this.returnType = output.getType();
        }
    }

    @Override
    public Type getType() {
        return returnType;
    }

    public List<DeclarationInstruction> getDeclarations() {
        return declarations;
    }

    public List<Variable> getInput() {
        return input;
    }

    public Variable getOutput() {
        return output;
    }

    public String getName() {
        return name;
    }

    @Override
    public void addDeclarationCommand(Variable variable) {
        DeclarationInstruction declaration = new DeclarationInstruction(variable);
        if (!isInputOrOutput(variable)) {
            declarations.add(declaration);
        }
    }

    private boolean isInputOrOutput(Variable variable) {
        String name = variable.getName();

        if (output != null && name.equals(output.getName())) {
            return true;
        }
        for (Variable value : input) {
            if (name.equals(value.getName())) {
                return true;
            }
        }
        return false;
    }

    public boolean isOutputAssigned() {
        if (output == null) {
            return true;
        }
        for (Instruction instruction : instructions) {
            if (instruction instanceof AssignmentInstruction) {
                if ( ((AssignmentInstruction) instruction).getVariable().getName().equals(output.getName()) ) {
                    return true;
                }
            }
        }

        return false;
    }
}

File: src/main/kotlin/br/com/pnp/model/construct/Procedure.kt
package br.com.pnp.model.construct

import br.com.pnp.model.construct.type.Type
import br.com.pnp.model.construct.type.primitive.PrimitiveType
import br.com.pnp.model.expression.Expression
import br.com.pnp.model.instruction.AssignmentInstruction
import br.com.pnp.model.instruction.DeclarationInstruction

class Procedure(val name: String): Block(), Construct, Expression {
    override val type: Type
        get() = output?.type ?: PrimitiveType.none

    private val declarationsList: MutableList<DeclarationInstruction> = ArrayList()
    val declarations get() = declarationsList.toList()

    var input: List<Variable> = ArrayList()
    var output: Variable? = null

    override fun addDeclarationInstruction(variable: Variable) {
        if (!isInputOrOutput(variable)) {
            declarationsList.add(DeclarationInstruction(variable))
        }
    }

    private fun isInputOrOutput(variable: Variable): Boolean {
        if (output == variable) {
            return true
        }
        if (input.find { it == variable } != null) {
            return true
        }
        return false
    }

    fun isOutputAssigned(): Boolean {
        return output?.let { variable ->
            val isAssigned: Boolean
            val assignment = instructions.filterIsInstance<AssignmentInstruction>()
                    .find { it.variable == variable }

            isAssigned = assignment != null
            // TODO: search inside the statements if there is no assignment in the procedure main block
            isAssigned
        } ?: true
    }
}